tests/queryparams/query_params_test.go

```diff
@@ -0,0 +1,240 @@
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		want := url.Values{"vegetable": []string{"ăȚâȘî"}}
+		got := req.URL.Query()
+		if diff := cmp.Diff(want, got); diff != "" {
+			t.Errorf("req.URL.Query(): got %v, want %v, diff (-want +got):\n%s", got, want, diff)
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(status200OK)) {
+		t.Errorf("response status: got %s, want %s", resp, status200OK)
+	}
+
+	req = []byte("GET /?ăȚâȘî=vegetable HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"\r\n")
+	resp, err = requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		want := url.Values{"ăȚâȘî": []string{"vegetable"}}
+		got := req.URL.Query()
+		if diff := cmp.Diff(want, got); diff != "" {
+			t.Errorf("req.URL.Query(): got %v, want %v, diff (-want +got):\n%s", got, want, diff)
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(status200OK)) {
+		t.Errorf("response status: got %s, want %s", resp, status200OK)
+	}
+}
+
+// Tests whether passing invalid Unicode in both key and value will result in a 400 Bad Request response
+func TestQueryParametersInvalidUnicodes(t *testing.T) {
+
+	tests := []struct {
+		name string
+		req  []byte
+	}{
+		{
+			name: "Invalid Key",
+			req: []byte("GET /?\x0F=tomato&Vegetable=potato HTTP/1.1\r\n" + "Host: localhost:8080\r\n" +
+				"\r\n"),
+		},
+		{
+			name: "Invalid value",
+			req: []byte("GET /?vegetable=\x0F&Vegetable=potato HTTP/1.1\r\n" + "Host: localhost:8080\r\n" +
+				"\r\n"),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			resp, err := requesttesting.MakeRequest(context.Background(), test.req, func(req *http.Request) {
+				t.Error("expected handler not to be called.")
+			})
+			if err != nil {
+				t.Fatalf("MakeRequest(): got err %v, want nil", err)
+			}
+			if !bytes.HasPrefix(resp, []byte(status400BadReq)) {
+				t.Errorf("response status: got %s, want %s", resp, status400BadReq)
+			}
+		})
+	}
+
+}
+
+// Verifies behaviour of query parameter parser when passing malformed keys or
+// values, by breaking URL percent-encoding. Percent-encoding is used to encode
+// special characters in URL
+func TestQueryParametersBreakUrlEncoding(t *testing.T) {
+	tests := []struct {
+		name string
+		req  []byte
+	}{
+		{
+			name: "Broken Key",
+			req: []byte("GET /?vegetable%=tomato&Vegetable=potato HTTP/1.1\r\n" + "Host: localhost:8080\r\n" +
+				"\r\n"),
+		},
+		{
+			name: "Broken value",
+			req: []byte("GET /?vegetable=tomato%&Vegetable=potato HTTP/1.1\r\n" + "Host: localhost:8080\r\n" +
+				"\r\n"),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			resp, err := requesttesting.MakeRequest(context.Background(), test.req, func(req *http.Request) {
+				var want []string = nil
```

Comment:
nit: `var want []string` is enough. Go has sane zero values.

Reply:
Resolved in [4504136bee3f68b1e4b142e96c26e322208484e7](url)